qualtran/bloqs/data_loading/qroam_clean.py

```diff
@@ -0,0 +1,389 @@
+class QROAMCleanAdjoint(QROMBase, GateWithRegisters):  # type: ignore[misc]
+    log_block_sizes: Tuple[SymbolicInt, ...] = attrs.field(
+        converter=lambda x: tuple(x.tolist() if isinstance(x, np.ndarray) else x)
+    )
+
+    @cached_property
+    def _target_reg_side(self) -> Side:
+        return Side.LEFT
+
+    @classmethod
+    def build_from_data(
+        cls: Type['QROAMCleanAdjoint'],
+        *data: ArrayLike,
+        target_bitsizes: Optional[Union[SymbolicInt, Tuple[SymbolicInt, ...]]] = None,
+        target_shapes: Tuple[Tuple[SymbolicInt, ...], ...] = (),
+        num_controls: SymbolicInt = 0,
+        log_block_sizes: Optional[Union[SymbolicInt, Tuple[SymbolicInt, ...]]] = None,
+    ) -> 'QROAMCleanAdjoint':
+        qroam: 'QROAMCleanAdjoint' = cls._build_from_data(
+            *data,
+            target_bitsizes=target_bitsizes,
+            target_shapes=target_shapes,
+            num_controls=num_controls,
+        )
+        return qroam.with_log_block_sizes(log_block_sizes)
+
+    @classmethod
+    def build_from_bitsize(
+        cls: Type['QROAMCleanAdjoint'],
+        data_len_or_shape: Union[SymbolicInt, Tuple[SymbolicInt, ...]],
+        target_bitsizes: Union[SymbolicInt, Tuple[SymbolicInt, ...]],
+        *,
+        target_shapes: Tuple[Tuple[SymbolicInt, ...], ...] = (),
+        selection_bitsizes: Tuple[SymbolicInt, ...] = (),
+        num_controls: SymbolicInt = 0,
+        log_block_sizes: Optional[Union[SymbolicInt, Tuple[SymbolicInt, ...]]] = None,
+    ) -> 'QROAMCleanAdjoint':
+        qroam: 'QROAMCleanAdjoint' = cls._build_from_bitsize(
+            data_len_or_shape,
+            target_bitsizes,
+            selection_bitsizes=selection_bitsizes,
+            target_shapes=target_shapes,
+            num_controls=num_controls,
+        )
+        return qroam.with_log_block_sizes(log_block_sizes=log_block_sizes)
+
+    @log_block_sizes.default
+    def _default_log_block_sizes(self) -> Tuple[SymbolicInt, ...]:
+        target_bitsize = sum(self.target_bitsizes) * sum(
+            prod(shape) for shape in self.target_shapes
+        )
+        return tuple(
+            get_optimal_log_block_size_clean_ancilla(ilen, target_bitsize, adjoint=True)
+            for ilen in self.data_shape
+        )
+
+    def with_log_block_sizes(
+        self, log_block_sizes: Optional[Union[SymbolicInt, Tuple[SymbolicInt, ...]]] = None
+    ) -> 'QROAMCleanAdjoint':
+        if log_block_sizes is None:
+            return self
+        if isinstance(log_block_sizes, (int, sympy.Basic, numbers.Number)):
+            log_block_sizes = (log_block_sizes,)
+        if not is_symbolic(*log_block_sizes):
+            assert all(1 <= 2**bs <= ilen for bs, ilen in zip(log_block_sizes, self.data_shape))
+        return attrs.evolve(self, log_block_sizes=log_block_sizes)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        X = prod([2**k for k in self.log_block_sizes])
```

Comment:
not sure what our policy is on pep8 variable names I think `data_size` etc is a bit more readable though

Reply:
will update.

Reply:
done